forte/mrdsrg-spin-integrated/dsrg_mrpt2_grad/dsrg_mrpt2_deriv_solve_zvector.cc

```diff
@@ -0,0 +1,1504 @@
+        D_mo["wz"] -= B["g,a1,z"] * B["gvw"] * Gamma1_["wv"] * I["z,a1"];
+    } else {
+        // VIRTUAL-CORE
+        D_mo["em"] -= V["m1,e,e1,m"] * I["e1,e"] * I["m1,m"];
+        D_mo["em"] -= V["e1,e,m1,m"] * I["e1,e"] * I["m1,m"];
+
+        // CORE-ACTIVE
+        D_mo["mw"] += V["u,m,n1,w"] * I["m,n1"] * I["wu"];
+        D_mo["mw"] -= Gamma1_["wv"] * V["v,m,n1,w"] * I["m,n1"];
+        D_mo["mw"] -= V["w,m,n1,v"] * Gamma1_["wv"] * I["m,n1"];
+        D_mo["mw"] += V["m,y,n1,v"] * Gamma2_["u,v,w,y"] * I["m,n1"] * I["wu"];
+        D_mo["mw"] += V["m,Y,n1,V"] * Gamma2_["u,V,w,Y"] * I["m,n1"] * I["wu"];
+
+        D_mo["mw"] -= V["w,v,n1,m"] * Gamma1_["wv"] * I["m,n1"];  
+        D_mo["mw"] += 0.5 * V["x,y,n1,m"] * Gamma2_["u,w,x,y"] * I["m,n1"] * I["wu"];
+        D_mo["mw"] -= 0.5 * V["xywv"] * Gamma2_["wvxy"] * one_vec["m"];
+        D_mo["mw"] -= V["xYwV"] * Gamma2_["wVxY"] * one_vec["m"];
+        D_mo["mw"] += V["u,w,n1,m"] * I["m,n1"] * I["wu"];  
+        D_mo["mw"] -= Gamma1_["wv"] * V["v,w,n1,m"] * I["m,n1"];
+
+        // VIRTUAL-ACTIVE
+        D_mo["ew"] += 0.5 * V["xywv"] * Gamma2_["wvxy"] * one_vec["e"];
+        D_mo["ew"] += V["xYwV"] * Gamma2_["wVxY"] * one_vec["e"];
+        D_mo["ew"] -= 0.5 * V["e1,e,x,y"] * Gamma2_["u,w,x,y"] * I["e,e1"] * I["uw"];
+
+        D_mo["ew"] -= V["e,y,e1,v"] * Gamma2_["u,v,w,y"] * I["e,e1"] * I["uw"];
+        D_mo["ew"] -= V["e,Y,e1,V"] * Gamma2_["u,V,w,Y"] * I["e,e1"] * I["uw"];
+
+        // ACTIVE-ACTIVE
+        D_mo["wz"] -= V["z,v,u1,w"] * Gamma1_["zv"] * I["u1,w"];
+        D_mo["wz"] += V["a1,v,w,z"] * Gamma1_["wv"] * I["z,a1"];
+    }
+
+    for (const std::string& row : {"vc", "ca", "va", "aa"}) {
+        int idx1 = block_dim[row];
+        int pre1 = preidx[row];
+        if (row != "aa") {
+            D_mo.block(row).iterate([&](const std::vector<size_t>& i, double& value) {
+                if (std::fabs(value) > err) {
+                    int index = pre1 + i[0] * idx1 + i[1];
+                    D.at(index) = 1.0 / value;
+                }
+            });
+        } else {
+            D_mo.block(row).iterate([&](const std::vector<size_t>& i, double& value) {
+                if (std::fabs(value) > err) { 
+                    if (i[0] > i[1]) {
+                        int index = pre1 + i[0] * (i[0] - 1) / 2 + i[1];
+                        D.at(index)  = 1.0 / value;
+                    }
+                }
+            });
+        }
+    }
+
+    // Attention :  d_ci are the approximate preconditioner components for the CI part
+    double d_ci = 0.0;
+    d_ci += 2.0 * H["mn"] * I["mn"];
+    d_ci += V_pmqm["m,m1"] * I["m,m1"];
+    d_ci -= Eref_ - Enuc_ - Efrzc_;
+
+    if (std::fabs(d_ci) > err) {
+        double value = 1.0 / d_ci;
+        int idx = preidx["ci"];
+        for (int i = 0; i < ndets; ++i) {    
+            D.at(idx + i) = value;
+        }
+    }
+}
+
+void DSRG_MRPT2::gmres_solver(std::vector<double> & x_new) {
```

Comment:
I think the `gmres_solver` should be moved to `helpers`. But this can be done afterwards.

Reply:
The Ax=b is specifically defined in this gmres_solver